src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -1853,20 +1856,24 @@ CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptMessage(const ByteSpan
                                                                   MutableByteSpan & plaintext) const
 {
     uint8_t * output = plaintext.data();
-    return Crypto::AES_CCM_decrypt(ciphertext.data(), ciphertext.size(), aad.data(), aad.size(), mic.data(), mic.size(), mKeyValue,
-                                   Crypto::kAES_CCM128_Key_Length, nonce.data(), nonce.size(), output);
+    return Crypto::AES_CCM_decrypt(ciphertext.data(), ciphertext.size(), aad.data(), aad.size(), mic.data(), mic.size(),
+                                   mEncryptionKey, Crypto::kAES_CCM128_Key_Length, nonce.data(), nonce.size(), output);
 }
 
-CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::EncryptPrivacy(MutableByteSpan & header, uint16_t session_id,
-                                                                  const ByteSpan & payload, const ByteSpan & mic) const
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::PrivacyEncrypt(const ByteSpan & input, const ByteSpan & nonce,
+                                                                  MutableByteSpan & output) const
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    // TODO(#21559): Add AES_CTR_encrypt and use that
+    return Crypto::AES_CCM_encrypt(input.data(), input.size(), nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
+                                   nonce.data(), nonce.size(), output.data(), nullptr, 0);
 }
 
-CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptPrivacy(MutableByteSpan & header, uint16_t session_id,
-                                                                  const ByteSpan & payload, const ByteSpan & mic) const
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::PrivacyDecrypt(const ByteSpan & input, const ByteSpan & nonce,
+                                                                  MutableByteSpan & output) const
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    // TODO(#21559): Add AES_CTR_decrypt and use that
```

Comment:
If this is not used yet, why do we cause anyone to have to pay the RAM cost for this feature?

Reply:
v1.0 nodes will need to be able to parse v1.0 message format.  That includes messages with privacy, MX, and / or SX enabled.

Reply:
Each privacy key is re-derived from the encryption_key in `KeySetData::Deserialize` to avoid any growth of `PersistentData`.